staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_websocket.go

```diff
@@ -0,0 +1,284 @@
+			return
+		}
+		if messageType != gwebsocket.BinaryMessage {
+			c.closeAllStreamReaders(fmt.Errorf("unexpected message type: %d", messageType))
+			return
+		}
+		// it's ok to read just a single byte because the underlying library wraps the actual connection with
+		// a buffered reader anyway.
+		_, err = io.ReadFull(r, readBuffer[:1])
+		if err != nil {
+			c.closeAllStreamReaders(fmt.Errorf("read stream id: %w", err))
+			return
+		}
+		streamID := readBuffer[0]
+		s := c.streams[streamID]
+		if s == nil {
+			klog.Errorf("Unknown stream id %d, discarding message", streamID)
+			continue
+		}
+		for {
+			nr, errRead := r.Read(readBuffer)
+			if nr > 0 {
+				_, errWrite := s.writePipe.Write(readBuffer[0:nr])
+				if errWrite != nil {
+					// Pipe must have been closed by the stream user. Nothing to do, discard the message.
+					break
+				}
+			}
+			if errRead != nil {
+				if errRead == io.EOF {
+					break
+				}
+				c.closeAllStreamReaders(fmt.Errorf("read message: %w", err))
+				return
+			}
+		}
+	}
+}
+
+// closeAllStreamReaders closes readers in all streams.
+// This unblocks all stream.Read() calls.
+func (c *wsStreamCreator) closeAllStreamReaders(err error) {
+	for _, s := range c.streams {
+		// Closing writePipe unblocks all readPipe.Read() callers and prevents any future writes.
+		_ = s.writePipe.CloseWithError(err)
+	}
+}
+
+type stream struct {
+	headers   http.Header
+	readPipe  *io.PipeReader
+	writePipe *io.PipeWriter
+	// conn is used for writing directly into the connection.
+	// Is nil after Close() / Reset() to prevent future writes.
+	conn *gwebsocket.Conn
+	// connMu protects conn against concurrent write operations. There must be a single writer and a single reader only.
+	// The mutex is shared across all streams because the underlying connection is shared.
+	connMu *sync.Mutex
+	id     byte
+}
+
+func (s *stream) Read(p []byte) (n int, err error) {
+	return s.readPipe.Read(p)
+}
+
+// Write writes directly to the underlying WebSocket connection.
+func (s *stream) Write(p []byte) (n int, err error) {
+	s.connMu.Lock()
+	defer s.connMu.Unlock()
+	if s.conn == nil {
+		return 0, fmt.Errorf("write on closed stream %d", s.id)
+	}
+	// TODO s.conn.SetWriteDeadline()
+	// Message writer buffers the message data, so we don't need to do that ourselves.
+	// Just write id and the data as two separate writes to avoid allocating an intermediate buffer.
+	w, err := s.conn.NextWriter(gwebsocket.BinaryMessage)
+	if err != nil {
+		return 0, err
+	}
+	_, err = w.Write([]byte{s.id})
+	if err != nil {
+		_ = w.Close()
+		return 0, err
+	}
+	n, err = w.Write(p)
+	if err != nil {
+		_ = w.Close()
+		return n, err
+	}
+	return n, w.Close()
+}
+
+// Close half-closes the stream, indicating this side is finished with the stream.
+func (s *stream) Close() error {
+	s.connMu.Lock()
+	defer s.connMu.Unlock()
+	s.conn = nil
+	// TODO See https://github.com/kubernetes/kubernetes/issues/89899#issuecomment-1132502190.
```

Comment:
We need to resolve https://github.com/kubernetes/kubernetes/issues/89899, otherwise this PR will cause a regression for those cases where it's important to be able to close STDIN. It'd be a regression because it works now via SPDY but will not with WebSocket API as it is today.